test/e2e/upgrade_hypershift_operator_test.go

```diff
@@ -0,0 +1,217 @@
+		if err != nil {
+			t.Fatal("Failed to install initial HyperShift Operator")
+		}
+
+	})).To(gomega.BeTrue(), "Installing initial HyperShift Operator should succeed")
+
+	t.Log("Executing upgrade test")
+	e2eutil.NewHypershiftTest(t, ctx, func(t *testing.T, g gomega.Gomega, mc crclient.Client, hc *hyperv1.HostedCluster) {
+		t.Logf("HostedCluster %s created", crclient.ObjectKeyFromObject(hc))
+		// Sanity check the cluster by waiting for the nodes to report ready
+		t.Log("Waiting for nodes to report ready")
+		_ = e2eutil.WaitForGuestClient(t, ctx, mc, hc)
+
+		t.Log("Nodes are ready")
+		mgmtClient = mc
+		hostedCluster = hc
+
+		g.Expect(t.Run("Calculate HyperShift Operator upgrade invariants", func(t *testing.T) {
+			t.Log("Calculating HyperShift Operator upgrade invariants")
+			g := gomega.NewWithT(t)
+
+			// Get the newly created default NodePool
+			nodepools := &hyperv1.NodePoolList{}
+			err = mgmtClient.List(ctx, nodepools, crclient.InNamespace(hostedCluster.Namespace))
+
+			g.Expect(err).ToNot(gomega.HaveOccurred(), "Listing nodepools in namespace %s shouldn't return errors",
+				hostedCluster.Namespace)
+			g.Expect(nodepools.Items).ToNot(gomega.BeEmpty(), "Should find NodePools in namespace %s",
+				hostedCluster.Namespace)
+
+			nodePoolsMap = make(map[string]*hyperv1.NodePool, len(nodepools.Items))
+			t.Logf("Found %d NodePools", len(nodepools.Items))
+			for i := range nodepools.Items {
+				t.Logf("Found NodePool %s", nodepools.Items[i].Name)
+				nodePoolsMap[nodepools.Items[i].Name] = &nodepools.Items[i]
+			}
+
+			// Get the MachineDeployments
+			machineDeployments := &v1beta1.MachineDeploymentList{}
+			hcpNameSpace = manifests.HostedControlPlaneNamespace(hostedCluster.Namespace, hostedCluster.Name)
+
+			err = mgmtClient.List(ctx, machineDeployments, crclient.InNamespace(hcpNameSpace))
+
+			g.Expect(err).ToNot(gomega.HaveOccurred(),
+				"Listing MachineDeployments in namespace %s shouldn't return errors", hcpNameSpace)
+			g.Expect(machineDeployments.Items).ToNot(gomega.BeEmpty(),
+				"Should find MachineDeployments in namespace %s", hcpNameSpace)
+			g.Expect(len(machineDeployments.Items)).To(gomega.BeEquivalentTo(len(nodepools.Items)),
+				"Number of MachineDeployments and NodePools should match")
+
+			machineDeploymentMap = make(map[string]*v1beta1.MachineDeployment)
+			t.Logf("Found %d MachineDeployments", len(machineDeployments.Items))
+			for i := range machineDeployments.Items {
+				t.Logf("Found MachineDeployment %s", machineDeployments.Items[i].Name)
+				machineDeploymentMap[machineDeployments.Items[i].Name] = &machineDeployments.Items[i]
+			}
+
+		})).To(gomega.BeTrue(), "Calculating HyperShift Operator upgrade invariants should succeed")
+
+		g.Expect(t.Run("Upgrade HyperShift Operator", func(t *testing.T) {
+			t.Logf("Upgrading HyperShift Operator to image %s", hyperShiftOperatorLatestImage)
+			globalOpts.HyperShiftOperatorLatestImage = hyperShiftOperatorLatestImage
```

Comment:
how do we know this is actually upgrading? can we print the current running image?

Reply:
good point 👍🏽

Reply:
Done

Reply:
sorry where is this done? I still see the same log sentence here
`t.Logf("Upgrading HyperShift Operator to image %s", hyperShiftOperatorLatestImage)`

Reply:
After each installation / upgrade of the HO, I get and verify the image. The few lines after the initial installation and the upgrade is done 

https://github.com/openshift/hypershift/pull/4999/files#diff-421ca9e9fe9c0358db093bc1035943aa8b1e1870b99f30a7480d8a28a885e65cR169-R171